qualtran/bloqs/arithmetic/controlled_addition.py

```diff
@@ -0,0 +1,240 @@
+            return
+        else:
+            if depth < 1:
+                raise ValueError(f"{depth=} is not a positive integer")
+            if depth < len(inp):
+                yield CNOT().on(anc[depth - 1], inp[depth])
+                control = inp[depth]
+            else:
+                # If inp[depth] doesn't exist, we treat it as a |0>,
+                # and therefore applying CNOT().on(anc[depth - 1], inp[depth])
+                # essentially "copies" anc[depth - 1] into inp[depth]
+                # in the classical basis. So therefore, on future operations,
+                # we can use anc[depth - 1] in its place.
+                control = anc[depth - 1]
+            yield CNOT().on(anc[depth - 1], out[depth])
+            yield And().on(control, out[depth], anc[depth])
+            yield CNOT().on(anc[depth - 1], anc[depth])
+            yield from self._left_building_block(inp, out, anc, depth + 1)
+
+    def _right_building_block(self, inp, out, anc, control, depth):
+        if depth == 0:
+            return
+        yield CNOT().on(anc[depth - 1], anc[depth])
+        if depth < len(inp):
+            yield And().adjoint().on(inp[depth], out[depth], anc[depth])
+            yield MultiControlPauli((1, 1), cirq.X).on(control, inp[depth], out[depth])
+            yield CNOT().on(anc[depth - 1], inp[depth])
+        else:
+            yield And().adjoint().on(anc[depth - 1], out[depth], anc[depth])
+            yield MultiControlPauli((1, 1), cirq.X).on(control, anc[depth - 1], out[depth])
+        yield CNOT().on(anc[depth - 1], out[depth])
+        yield from self._right_building_block(inp, out, anc, control, depth - 1)
+
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        # reverse the order of qubits for big endian-ness.
+        input_bits = quregs['a'][::-1]
+        output_bits = quregs['b'][::-1]
+        ancillas = context.qubit_manager.qalloc(self.b_dtype.bitsize - 1)[::-1]
+        control = quregs['ctrl'][0]
+        if self.controlled == 0:
+            yield cirq.X(control)
+        # Start off the addition by anding into the ancilla
+        yield And().on(input_bits[0], output_bits[0], ancillas[0])
+        # Left part of Fig.4
+        yield from self._left_building_block(input_bits, output_bits, ancillas, 1)
+        yield CNOT().on(ancillas[-1], output_bits[-1])
+        if len(input_bits) == len(output_bits):
+            yield MultiControlPauli((1, 1), cirq.X).on(control, input_bits[-1], output_bits[-1])
+            yield CNOT().on(ancillas[-1], output_bits[-1])
+        # right part of Fig.4
+        yield from self._right_building_block(
+            input_bits, output_bits, ancillas, control, self.b_dtype.bitsize - 2
+        )
+        yield And().adjoint().on(input_bits[0], output_bits[0], ancillas[0])
+        yield MultiControlPauli((1, 1), cirq.X).on(control, input_bits[0], output_bits[0])
```

Comment:
Just a Toffoli right? Could this be another And gate?

Reply:
I used MultiControlPauli because this dynamically uses the available free ancillas. Or was that not what you were asking?